core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -235,16 +236,18 @@ public String toString() {
   private final List<SortOrder> sortOrders;
   private final Map<String, String> properties;
   private final long currentSnapshotId;
-  private final List<Snapshot> snapshots;
-  private final Map<Long, Snapshot> snapshotsById;
   private final Map<Integer, Schema> schemasById;
   private final Map<Integer, PartitionSpec> specsById;
   private final Map<Integer, SortOrder> sortOrdersById;
   private final List<HistoryEntry> snapshotLog;
   private final List<MetadataLogEntry> previousFiles;
-  private final Map<String, SnapshotRef> refs;
   private final List<StatisticsFile> statisticsFiles;
   private final List<MetadataUpdate> changes;
+  private final SerializableSupplier<List<Snapshot>> snapshotsSupplier;
+  private List<Snapshot> snapshots;
```

Comment:
I think these variables should also be `transient` since they are modified by `ensureSnapshotsLoaded` and that could be called by multiple threads.

Reply:
Ok, I've updated to be `volatile` and `ensureSnapshotsLoaded` should be synchronized to make it thread safe.